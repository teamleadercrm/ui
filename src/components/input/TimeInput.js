import React, { forwardRef, PureComponent } from 'react';
import InputMask from 'react-input-mask';
import Icon from '../icon';
import SingleLineInputBase from './SingleLineInputBase';
import { IconTimerSmallOutline } from '@teamleader/ui-icons';

const isValidTime = (input) => /([0-1][0-9]|[2][0-3]):([0-5][0-9])/.test(input);
class TimeInput extends PureComponent {
  beforeMaskedValueChange = ({ value: newValue, selection }, { value: oldValue }) => {
    if (!isValidTime(newValue)) {
      return { value: oldValue, selection };
    }
    return { value: newValue, selection };
  };

  render() {
    const { disabled, readOnly, forwardedRef } = this.props;
    return (
      <InputMask {...this.props} mask="99:99" maskChar="0" beforeMaskedValueChange={this.beforeMaskedValueChange}>
        {(inputProps) => (
          <SingleLineInputBase
            {...inputProps}
            ref={forwardedRef}
            autoComplete="off"
            readOnly={readOnly}
            disabled={disabled}
            prefix={
              <Icon color="neutral" tint="darkest">
                <IconTimerSmallOutline />
              </Icon>
            }
          />
        )}
      </InputMask>
    );
  }
}

/** @type {React.ComponentType<any>} */
const ForwardedTimeInput = forwardRef((props, ref) => <TimeInput {...props} forwardedRef={ref} />);

ForwardedTimeInput.displayName = 'TimeInput';

export default ForwardedTimeInput;
