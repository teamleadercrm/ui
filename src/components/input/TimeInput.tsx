import React, { forwardRef } from 'react';
import ReactInputMask, { InputState } from 'react-input-mask';
import Icon from '../icon';
import SingleLineInputBase from './SingleLineInputBase';
import { IconTimerSmallOutline } from '@teamleader/ui-icons';

const isValidTime = (input: string) => /([0-1][0-9]|[2][0-3]):([0-5][0-9])/.test(input);

interface TimeInputProps {
  /** Boolean indicating whether the input should render as disabled. */
  disabled?: boolean;
  /** Boolean indicating whether the input should render as read only. */
  readOnly?: boolean;
}

const TimeInput = forwardRef((props: TimeInputProps, ref) => {
  const { readOnly, disabled } = props;

  const beforeMaskedValueChange = (
    { value: newValue, selection }: InputState,
    { value: oldValue }: InputState,
  ): InputState => {
    if (!isValidTime(newValue)) {
      return { value: oldValue, selection };
    }

    return { value: newValue, selection };
  };

  return (
    <ReactInputMask {...props} mask="99:99" maskChar="0" beforeMaskedValueChange={beforeMaskedValueChange}>
      {(inputProps: TimeInputProps) => (
        <SingleLineInputBase
          {...inputProps}
          ref={ref}
          autoComplete="off"
          readOnly={readOnly}
          disabled={disabled}
          prefix={
            <Icon color="neutral" tint="darkest">
              <IconTimerSmallOutline />
            </Icon>
          }
        />
      )}
    </ReactInputMask>
  );
});

TimeInput.displayName = 'TimeInput';

export default TimeInput;
