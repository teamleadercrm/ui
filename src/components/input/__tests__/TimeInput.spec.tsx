import React from 'react';
import { render } from '@testing-library/react';
import TimeInput from '../TimeInput';

describe('Component - TimeInput', () => {
  it('renders', async () => {
    const { asFragment } = render(<TimeInput />);
    expect(asFragment()).toMatchSnapshot();
  });
});
