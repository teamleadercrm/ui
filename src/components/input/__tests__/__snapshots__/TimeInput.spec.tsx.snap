// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - TimeInput renders 1`] = `
<DocumentFragment>
  <div
    class="box wrapper"
  >
    <div
      class="input-wrapper"
    >
      <div
        class="input-inner-wrapper"
      >
        <div
          class="prefix-wrapper"
        >
          <span
            class="box align-items-center display-flex neutral darkest"
            data-teamleader-ui="icon"
          >
            <svg
              fill="currentColor"
              height="14"
              preserveAspectRatio="xMidYMid meet"
              style="opacity: 0.84; vertical-align: middle;"
              viewBox="0 0 14 14"
              width="14"
            >
              <path
                d="M7 3a4 4 0 10.001 8.001A4 4 0 007 3zm2-1.659A6.002 6.002 0 017 13 6 6 0 015 1.342V1a1 1 0 011-1h2a1 1 0 011 1v.341zM7 4a1 1 0 011 1v2a1 1 0 01-2 0V5a1 1 0 011-1z"
                fill="currentColor"
              />
            </svg>
          </span>
        </div>
        <input
          autocomplete="off"
          class="input is-medium"
          data-teamleader-ui="time-input"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
